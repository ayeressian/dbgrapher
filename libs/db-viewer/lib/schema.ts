import type Point from "./point";

export interface FkSchema {
  table: string;
  column: string;
}
export interface ColumnNoneFkSchema {
  name: string;
  type: string;
  pk?: boolean;
  uq?: boolean;
  nn?: boolean;
}

export declare interface ColumnFkSchema
  extends Omit<ColumnNoneFkSchema, "type"> {
  fk: FkSchema;
}

export type ColumnSchema = ColumnFkSchema | ColumnNoneFkSchema;

export interface TableSchema {
  name: string;
  pos: Point;
  columns: ColumnSchema[];
}

export interface AnnotationSchema {
  pos: Point;
  title?: string;
  text?: string;
  color?: string;
  width: number;
  height: number;
}

export interface Schema {
  tables: TableSchema[];
  annotations?: AnnotationSchema[];
  arrangement?: TableArrang;
  viewport?: Viewport;
  viewWidth?: number;
  viewHeight?: number;
}

export const enum Viewport {
  noChange = "noChange",
  centerByTablesWeight = "centerByTablesWeight",
  center = "center",
  centerByTables = "centerByTables",
}

export enum TableArrang {
  spiral = "spiral",
  default = "default",
}

export const isColumnFk = (column: ColumnSchema): column is ColumnFkSchema => {
  return (column as ColumnFkSchema).fk !== undefined;
};
