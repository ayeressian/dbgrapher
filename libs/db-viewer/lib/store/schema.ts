import { get, writable, type Readable } from "svelte/store";
import type { Schema } from "../schema";
import type { Store } from "./store";
import type { Viewport } from "./view";

export type RectSize = {
  width: number;
  height: number;
};

export class SchemaStore {
  private store: Store;
  schema = writable<Schema>({ tables: [] });
  // allTableSizesSetStore: Readable<boolean>;

  constructor(store: Store) {
    this.store = store;
  }

  getSchema(): Schema {
    const schema = get(this.schema);
    this.store.table.updateTablePoses(schema.tables);
    return schema;
  }

  setSchema(schema: Schema, viewport: Viewport = "center") {
    schema = JSON.parse(JSON.stringify(schema));
    this.store?.view.viewportStore.set(viewport);
    const tableStore = this.store.table;
    tableStore.tableSizes.clear();
    tableStore.tablePoses.clear();
    tableStore.tableToHighlight.clear();
    tableStore.tableFromHighlight.clear();
    schema.tables.forEach((table) => {
      const tableSizeWritable = writable<RectSize>();
      tableStore.tableSizes.set(table.name, tableSizeWritable);
      const tablePosWritable = writable(table.pos);
      tableStore.tablePoses.set(table.name, tablePosWritable);
      tablePosWritable.subscribe(this.store!.view.getUpdateViewBound());
      tableStore.tableToHighlight.set(table.name, writable(""));
      tableStore.tableFromHighlight.set(table.name, writable(""));
    });
    this.schema.set({tables: []});
    setTimeout(() => this.schema.set(schema));    

    // allTableSizesSetStore = derived(
    //   [...tableSizes.values()],
    //   (tableSizes) => {
    //     return tableSizes.every((tableSize) => {
    //       return tableSize.height > 0 && tableSize.width > 0;
    //     });
    //   }
    // );
  }
}
