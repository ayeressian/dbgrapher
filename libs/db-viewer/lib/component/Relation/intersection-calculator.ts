import type { RectSize } from "../../store/schema";
import { segmentIntersection } from "../../math-util";
import type Point from "../../point";

export default class IntersectionCalculation {
  private halfWidth: number;
  private halfHight: number;

  constructor(
    private tablePos: Point,
    tableSize: RectSize,
    private point1: Point,
    private point2: Point,
  ) {
    this.halfWidth = tableSize.width / 2;
    this.halfHight = tableSize.height / 2;
  }

  call() {
    return (
      this.intersectFromTop() ||
      this.intersectFromBottom() ||
      this.intersectFromLeft() ||
      this.intersectFromRight()
    );
  }

  private intersectFromRight() {
    return segmentIntersection(
      {
        x: this.tablePos.x - this.halfWidth,
        y: this.tablePos.y - this.halfHight,
      },
      {
        x: this.tablePos.x - this.halfWidth,
        y: this.tablePos.y + this.halfHight,
      },
      this.point1,
      this.point2
    );
  }

  private intersectFromLeft() {
    return segmentIntersection(
      {
        x: this.tablePos.x + this.halfWidth,
        y: this.tablePos.y - this.halfHight,
      },
      {
        x: this.tablePos.x + this.halfWidth,
        y: this.tablePos.y + this.halfHight,
      },
      this.point1,
      this.point2
    );
  }

  private intersectFromBottom() {
    return segmentIntersection(
      {
        x: this.tablePos.x - this.halfWidth,
        y: this.tablePos.y + this.halfHight,
      },
      {
        x: this.tablePos.x + this.halfWidth,
        y: this.tablePos.y + this.halfHight,
      },
      this.point1,
      this.point2
    );
  }

  private intersectFromTop() {
    return segmentIntersection(
      {
        x: this.tablePos.x - this.halfWidth,
        y: this.tablePos.y - this.halfHight,
      },
      {
        x: this.tablePos.x + this.halfWidth,
        y: this.tablePos.y - this.halfHight,
      },
      this.point1,
      this.point2
    );
  }
}
