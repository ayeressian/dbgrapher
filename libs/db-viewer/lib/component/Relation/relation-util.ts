import { rotateCord } from "../../math-util";
import type Point from "../../point";

export const RELATION_PADDING = 10;

export const oneEnd = (angle: number, end: Point): string => {
  const lineStart = rotateCord(
    {
      x: -RELATION_PADDING + end.x,
      y: -RELATION_PADDING / 2 + end.y,
    },
    angle,
    end
  );

  const lineEnd = rotateCord(
    {
      x: -RELATION_PADDING + end.x,
      y: RELATION_PADDING / 2 + end.y,
    },
    angle,
    end
  );

  return `M ${lineStart.x} ${lineStart.y} L ${lineEnd.x} ${lineEnd.y}`;
};

export const manyEnd = (angle: number, end: Point): string => {
  const lineStart = rotateCord(
    {
      x: -RELATION_PADDING + end.x,
      y: end.y,
    },
    angle,
    end
  );

  const firstLineEnd = rotateCord(
    {
      x: end.x,
      y: RELATION_PADDING / 2 + end.y,
    },
    angle,
    end
  );

  const secondLineEnd = rotateCord(
    {
      x: end.x,
      y: -RELATION_PADDING / 2 + end.y,
    },
    angle,
    end
  );

  return `M ${lineStart.x} ${lineStart.y} L ${firstLineEnd.x} ${firstLineEnd.y} M ${lineStart.x} ${lineStart.y} L ${secondLineEnd.x} ${secondLineEnd.y}`;
};

export const oneStart = (angle: number, start: Point): string => {
  const lineStart = rotateCord(
    {
      x: RELATION_PADDING + start.x,
      y: -RELATION_PADDING / 2 + start.y,
    },
    angle,
    start
  );

  const lineEnd = rotateCord(
    {
      x: RELATION_PADDING + start.x,
      y: RELATION_PADDING / 2 + start.y,
    },
    angle,
    start
  );

  return `M ${lineStart.x} ${lineStart.y} L ${lineEnd.x} ${lineEnd.y}`;
};

export const zeroStart = (angle: number, start: Point): string => {
  const circleCord = rotateCord(
    {
      x: RELATION_PADDING + start.x,
      y: start.y,
    },
    angle,
    start
  );

  const newStart = rotateCord(
    {
      x: RELATION_PADDING * 2,
      y: 0,
    },
    angle
  );

  start.x = newStart.x + start.x;
  start.y = newStart.y + start.y;

  return (
    `M ${circleCord.x - RELATION_PADDING} ${circleCord.y}` +
    ` a 1,1 0 1,0 ${RELATION_PADDING * 2},0 a 1,1 0 1,0 ${
      -RELATION_PADDING * 2
    },0`
  );
};
