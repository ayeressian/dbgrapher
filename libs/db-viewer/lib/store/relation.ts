import { isColumnFk } from "../schema";
import { derived, get, type Readable } from "svelte/store";
import type { Store } from "./store";

type Relation = {
  key: string;
  fromTable: string;
  toTable: string;
  fromColumn: string;
  toColumn: string;
  oneTo: boolean;
  toMany: boolean;
};

export class RelationStore {
  relations?: Readable<Relation[]>;

  constructor(private store: Store) {
    this.relations = derived(store.schema.schema, ($schema) => {
      const relations: Relation[] = [];
      let index = 0;
      $schema.tables.forEach((table) => {
        table.columns.forEach((column) => {
          if (isColumnFk(column)) {
            relations.push({
              key: `${table.name}_${column.name}_${column.fk.table}_${column.fk.column}`,
              toTable: table.name,
              fromTable: column.fk.table,
              toColumn: column.name,
              fromColumn: column.fk.column,
              oneTo: !!column.nn,
              toMany: !column.uq,
            });
            ++index;
          }
        });
      });
      return relations;
    });
  }

  getRelations(fromTable: string, toTable: string) {
    const realtionsVal = get(this.relations!);
    const r = realtionsVal.filter(
      (relation) =>
        relation.fromTable === fromTable && relation.toTable === toTable
    );
    return r;
  }
}
