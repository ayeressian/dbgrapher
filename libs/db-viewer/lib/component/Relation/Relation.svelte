<script lang="ts">
  import { rotateCord, segmentAngleRadRelativeToXAxis } from "../../math-util";
  import type Point from "../../point";

  import { manyEnd, oneEnd, oneStart, zeroStart } from "./relation-util";
  import IntersectionCalculation from "./intersection-calculator";
  import { getContext } from "svelte";
  import type { Store } from "../../store/store";
  import type { RelationData } from "../DBViewer/event";
  import type { RectSize } from "../../store/schema";

  export let fromTable: string;
  export let toTable: string;
  export let fromColumn: string;
  export let toColumn: string;
  export let oneTo: boolean;
  export let toMany: boolean;
  export let key: string;

  type EventCallback = (tableData: RelationData) => void;

  export let click: EventCallback;
  export let dblClick: EventCallback;
  export let contextMenu: EventCallback;

  const DISTANCE_BETWEEN_ADJASENT_RELATIONS = 20;

  const store = getContext<Store>("store");

  const fromTablePos = store.table.getTablePos(fromTable);
  const toTablePos = store.table.getTablePos(toTable);
  const fromTableSize = store.table.getTableSize(fromTable);
  const toTableSize = store.table.getTableSize(toTable);

  const adjacentRelations = store.relation.getRelations(fromTable, toTable);
  const adjacentOposRelations = store.relation.getRelations(toTable, fromTable);

  const indexFromAdjacent = adjacentRelations
    .sort((relation1, relation2) => relation1.key.localeCompare(relation2.key))
    .findIndex((relation) => relation.key === key);

  const totalRelationLength =
    adjacentRelations.length + adjacentOposRelations.length;

  const multipleSegmentLength =
    (totalRelationLength - 1) * DISTANCE_BETWEEN_ADJASENT_RELATIONS;

  let start: Point | null, end: Point | null;

  let startD = "";
  let endD = "";

  const drawStart = (angle: number, start: Point) => {
    startD = oneTo ? oneStart(angle, start) : zeroStart(angle, start);
  };

  const drawEnd = (angle: number, end: Point) => {
    endD = toMany ? manyEnd(angle, end) : oneEnd(angle, end);
  };

  const multipleSegmentPoints = (
    angle: number,
    fromTablePos: Point,
    toTablePos: Point
  ) => {
    const multipleSegmentAngle = angle + Math.PI / 2;
    let multipleSegmentPoint1Start = {
      x: fromTablePos.x - multipleSegmentLength / 2,
      y: fromTablePos.y,
    };
    let multipleSegmentPoint1End = {
      x: fromTablePos.x + multipleSegmentLength / 2,
      y: fromTablePos.y,
    };

    multipleSegmentPoint1Start = rotateCord(
      multipleSegmentPoint1Start,
      multipleSegmentAngle,
      fromTablePos
    );
    multipleSegmentPoint1End = rotateCord(
      multipleSegmentPoint1End,
      multipleSegmentAngle,
      fromTablePos
    );

    let multipleSegmentPoint2Start = {
      x: toTablePos.x - multipleSegmentLength / 2,
      y: toTablePos.y,
    };
    let multipleSegmentPoint2End = {
      x: toTablePos.x + multipleSegmentLength / 2,
      y: toTablePos.y,
    };

    multipleSegmentPoint2Start = rotateCord(
      multipleSegmentPoint2Start,
      multipleSegmentAngle,
      toTablePos
    );
    multipleSegmentPoint2End = rotateCord(
      multipleSegmentPoint2End,
      multipleSegmentAngle,
      toTablePos
    );

    return {
      multipleSegmentPoint1Start,
      multipleSegmentPoint1End,
      multipleSegmentPoint2Start,
      multipleSegmentPoint2End,
    };
  };

  const getMultiLineStart = (
    angle: number,
    fromTablePos: Point,
    toTablePos: Point
  ) => {
    const {
      multipleSegmentPoint1Start,
      multipleSegmentPoint1End,
      multipleSegmentPoint2Start,
      multipleSegmentPoint2End,
    } = multipleSegmentPoints(angle, fromTablePos, toTablePos);

    const deltaX = multipleSegmentPoint1End.x - multipleSegmentPoint1Start.x;
    const deltaY = multipleSegmentPoint1End.y - multipleSegmentPoint1Start.y;
    const lineStart = {
      x:
        multipleSegmentPoint1Start.x +
        (deltaX / (totalRelationLength - 1)) * indexFromAdjacent,
      y:
        multipleSegmentPoint1Start.y +
        (deltaY / (totalRelationLength - 1)) * indexFromAdjacent,
    };

    const deltaEndX = multipleSegmentPoint2End.x - multipleSegmentPoint2Start.x;
    const deltaEndY = multipleSegmentPoint2End.y - multipleSegmentPoint2Start.y;
    const lineEnd = {
      x:
        multipleSegmentPoint2Start.x +
        (deltaEndX / (totalRelationLength - 1)) * indexFromAdjacent,
      y:
        multipleSegmentPoint2Start.y +
        (deltaEndY / (totalRelationLength - 1)) * indexFromAdjacent,
    };

    return { lineStart, lineEnd };
  };

  const setStartAndEnd = (
    fromTableSize?: RectSize,
    toTableSize?: RectSize,
    fromTablePos?: Point,
    toTablePos?: Point
  ) => {
    if (
      fromTableSize?.height &&
      toTableSize?.height &&
      fromTableSize?.width &&
      toTableSize?.width &&
      fromTablePos &&
      toTablePos &&
      toTable !== fromTable
    ) {
      const angle = segmentAngleRadRelativeToXAxis(fromTablePos, toTablePos);
      let lineStart: Point;
      let lineEnd: Point;
      if (totalRelationLength === 1) {
        lineStart = fromTablePos;
        lineEnd = toTablePos;
      } else {
        ({ lineStart, lineEnd } = getMultiLineStart(
          angle,
          fromTablePos,
          toTablePos
        ));
      }

      start = new IntersectionCalculation(
        fromTablePos,
        fromTableSize,
        lineStart,
        lineEnd
      ).call();

      end = new IntersectionCalculation(
        toTablePos,
        toTableSize,
        lineStart,
        lineEnd
      ).call();

      if (start) {
        drawStart(angle, start);
      }
      if (end) {
        drawEnd(angle, end);
      }
    }
  };

  $: {
    setStartAndEnd($fromTableSize, $toTableSize, $fromTablePos, $toTablePos);
  }

  const onMouseEnter = () => {
    store.table.setTableFromHighlight(fromTable, fromColumn);
    store.table.setTableToHighlight(toTable, toColumn);
  };

  const onMouseLeave = () => {
    store.table.setTableFromHighlight(fromTable, "");
    store.table.setTableToHighlight(toTable, "");
  };

  const toHighlight = store.table.getTableFromHighlight(fromTable);

  const getRelationData = () => ({
    fromTable,
    toTable,
    fromColumn,
    toColumn,
  });

  const dblClickInternal = () => dblClick(getRelationData());
  const clickInternal = () => click(getRelationData());
  const contextMenuInternal = () => contextMenu(getRelationData());
</script>

{#if start != null && end != null}
  <path
    d="M {start.x} {start.y} L {end.x} {end.y} {startD} {endD}"
    class:highlight-path={$toHighlight === fromColumn}
  />
  <!-- svelte-ignore a11y-click-events-have-key-events -->
  <path
    on:dblclick={dblClickInternal}
    on:click={clickInternal}
    on:contextmenu={contextMenuInternal}
    on:mouseenter={onMouseEnter}
    on:mouseleave={onMouseLeave}
    d="M {start.x} {start.y} L {end.x} {end.y} {startD} {endD}"
    class="highlight"
  />
{/if}
