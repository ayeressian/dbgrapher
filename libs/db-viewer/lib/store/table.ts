import type Point from "../point";
import { get, type Writable } from "svelte/store";
import type { RectSize } from "./schema";
import type { TableSchema } from "lib/schema";

type TableSize = RectSize;

export class TableStore {
  tableSizes = new Map<string, Writable<TableSize>>();
  tablePoses = new Map<string, Writable<Point>>();
  tableFromHighlight = new Map<string, Writable<string>>();
  tableToHighlight = new Map<string, Writable<string>>();

  getTableSize(tableName: string) {
    return this.tableSizes.get(tableName);
  }
  getTablePos(tableName: string) {
    return this.tablePoses.get(tableName);
  }
  getTableFromHighlight(tableName: string) {
    return this.tableFromHighlight.get(tableName);
  }
  getTableToHighlight(tableName: string) {
    return this.tableToHighlight.get(tableName);
  }
  setTableFromHighlight(tableName: string, column: string) {
    this.tableFromHighlight.get(tableName)?.set(column);
  }
  setTableToHighlight(tableName: string, column: string) {
    this.tableToHighlight.get(tableName)?.set(column);
  }

  updateTablePoses(tableSchemas: TableSchema[]) {
    tableSchemas.forEach((tableSchema) => {
      const pos = this.getTablePos(tableSchema.name)!;
      tableSchema.pos = get(pos);
    });
  }
}
