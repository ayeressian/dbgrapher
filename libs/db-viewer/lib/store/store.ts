import { RelationStore } from "./relation";
import { SchemaStore } from "./schema";
import { TableStore } from "./table";
import { ViewStore } from "./view";

export class Store {
  schema = new SchemaStore(this);
  table = new TableStore();
  view = new ViewStore(this);
  relation = new RelationStore(this);
}
