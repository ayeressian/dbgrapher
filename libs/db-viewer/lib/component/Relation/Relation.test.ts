import { cleanup, render, type RenderResult } from "@testing-library/svelte";
import Relation from "./Relation.svelte";
import { setSchema } from "../../store/schema";
import school from "../../../src/school";
import { beforeEach, describe, expect, it } from "vitest";
import { getTableSize } from "lib/store/table";

const setTimeoutAsync = (timer = 0) =>
  new Promise((resolve) => setTimeout(resolve, timer));

describe(Relation.name, () => {
  let component: RenderResult;

  beforeEach(async () => {
    setSchema(school);
    getTableSize("school")?.set({
      width: 200,
      height: 300,
    });
		getTableSize("student")?.set({
      width: 200,
      height: 400,
    });
    cleanup();
    component = render(Relation, {
      fromTable: "school",
      toTable: "student",
      oneTo: true,
      toMany: false,
    });
  });
  it("should render properly", () => {
    expect(component).toMatchSnapshot();
  });
});
