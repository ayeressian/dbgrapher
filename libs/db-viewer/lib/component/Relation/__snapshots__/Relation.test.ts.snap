// Vitest Snapshot v1

exports[`Relation$ > should render properly 1`] = `
{
  "component": Relation$ {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        {
          "x": 622,
          "y": 526.0753880266075,
        },
        {
          "x": 371,
          "y": 440.9246119733925,
        },
        "M 610.9237853201372 527.5977113712331 L 614.1364125163818 518.1278102894927",
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        {
          "set": [Function],
          "subscribe": [Function],
          "update": [Function],
        },
        "school",
        "student",
        true,
        false,
        {
          "x": 271,
          "y": 407,
        },
        {
          "x": 722,
          "y": 560,
        },
        {
          "height": 400,
          "width": 200,
        },
        {
          "height": 300,
          "width": 200,
        },
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "fromTable": 7,
        "oneTo": 9,
        "toMany": 10,
        "toTable": 8,
      },
      "root": <div>
        <path
          d="M 622 526.0753880266075 L 371 440.9246119733925 M 610.9237853201372 527.5977113712331 L 614.1364125163818 518.1278102894927"
        />
         
        <path
          class="highlight"
          d="M 622 526.0753880266075 L 371 440.9246119733925 M 610.9237853201372 527.5977113712331 L 614.1364125163818 518.1278102894927"
        />
        
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <path
        d="M 622 526.0753880266075 L 371 440.9246119733925 M 610.9237853201372 527.5977113712331 L 614.1364125163818 518.1278102894927"
      />
       
      <path
        class="highlight"
        d="M 622 526.0753880266075 L 371 440.9246119733925 M 610.9237853201372 527.5977113712331 L 614.1364125163818 518.1278102894927"
      />
      
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
